### PR TITLE
DOC: Contributing guide docker expansion

### PR DESCRIPTION
**Pull request by fdanieluk**
Added instructions for creating development environment with docker on windows, mac os and unix.

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: doc/source/development/contributing.rst [hunk 1]
```diff
@@ -157,9 +157,56 @@ Instead of manually setting up a development environment, you can use Docker to
...
+
+See `official docker documentation <https://docs.docker.com/docker-for-mac/install/>`_ for details.
+
+**Test your Docker installation**
+
+Run the following command in your terminal::
+
+    docker run hello-world
+
+You should see a welcome message, your installation is complete! In case of problems refer to these pages with common issues on: `Windows <>https://docs.docker.com/docker-for-windows/troubleshoot/`_,
```
Comment (MarcoGorelli) [inline doc/source/development/contributing.rst:197]: Does the Windows hyperlink display correctly for you? (for me it doesn't)

File: doc/source/development/contributing.rst [hunk 2]
```diff
@@ -157,9 +157,56 @@ Instead of manually setting up a development environment, you can use Docker to
...
+See `official docker documentation <https://docs.docker.com/docker-for-mac/install/>`_ for details.
+
+**Test your Docker installation**
+
+Run the following command in your terminal::
+
+    docker run hello-world
+
+You should see a welcome message, your installation is complete! In case of problems refer to these pages with common issues on: `Windows <>https://docs.docker.com/docker-for-windows/troubleshoot/`_, 
+`MacOS <https://docs.docker.com/docker-for-mac/troubleshoot/>`_.
```
Comment (MarcoGorelli) [inline doc/source/development/contributing.rst:198]: Admittedly you've already linked to the Linux post-installation instructions, but is there a troubleshooting guide for Linux as well?

File: doc/source/development/contributing.rst [hunk 3]
```diff
@@ -157,9 +157,56 @@ Instead of manually setting up a development environment, you can use Docker to
...
+
+**Development inside Docker container**
+
+You can launch a remote session with Visual Studio Code, a popular free IDE, using the `.devcontainer.json` file in root directory. To do that you need to 
+Install `Visual Studio Code <https://code.visualstudio.com/insiders/>`_ and the `Remote Development extension pack <https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack>`_. Then follow these `steps <https://code.visualstudio.com/docs/remote/containers#_quick-start-try-a-dev-container>`_ to connect to remote session.
```
Comment (MarcoGorelli) [inline doc/source/development/contributing.rst:203]: Should the link be "Try a dev container" or https://code.visualstudio.com/docs/remote/containers#_quick-start-open-an-existing-folder-in-a-container ?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Great, thanks @fdaniuluk! I think this will make it more accessible for beginners

File: doc/source/development/contributing.rst (hunk 1 above)
Comment (fdanieluk) [reply doc/source/development/contributing.rst]: My bad, there's a typo, I commited a fix

File: doc/source/development/contributing.rst (hunk 2 above)
Comment (fdanieluk) [reply doc/source/development/contributing.rst:198]: On docker website they list Docker Engine (Unix) and Docker Desktop (Win and MacOS) separately. There is a section for troubleshooting for Docker Desktop, but there's no such section for Docker Engine, similarly there is no post-installation section for Docker Desktop

File: doc/source/development/contributing.rst (hunk 3 above)
Comment (fdanieluk) [reply doc/source/development/contributing.rst]: You're right, makes more sense. It's fixed now.

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Cool, thanks for updating - seems there's some linting issues now (trailing whitespaces), can you address those please?

**Review by MarcoGorelli — APPROVED** [review]
Thanks

EDIT: On a second thought, I'm wondering if the Windows section needs to be so long...will come back to this

**Comment by fdanieluk** [comment]
Sure, let me know how you’d like it to look. I can shorten it a bit if needed.

**Review by MarcoGorelli — COMMENTED** [review]
Great, thanks @fdanieluk for your openness to feedback :)

IMO what you've added is valuable because Docker is arguably the easiest way to get started, so it's good to make this section more accessible.

Here's what I think can be added to the version on `master`:
1. linking to the official Docker installation instructions, including (if applicable) post-installation instructions (which you've done :+1:) 
2. mentioning how to check the Docker installation is correct, and linking to troubleshooting steps (which you've also done :+1:)
3. mentioning how to develop using a Docker container without having to use vscode or PyCharm. I'm no Docker expert, but I think this would be:
```bash
docker build --tag pandas-dev:latest .
docker run -it --rm -v $PWD:/pandas-dev -w /pandas-dev pandas-dev:latest
```
4. mentioning how this can be done in PyCharm (which you've added :+1:)
5. mentioning how you will still need to rebuild the C extensions when you pull in new changes by doing
```bash
python setup.py --build_ext --inplace -j 4
```

The part I'm unsure of is that as the official docs already go through the details about WSL2, I wonder if it's necessary to repeat them here.

@yehoshuadimarsky @gfyoung @WillAyd any thoughts?

As an aside, we had a few people at PyData Amsterdam successfully set their environments up using Docker, so I definitely think it's been worthwhile to add this section!

**Comment by gfyoung** [comment]
How much documentation do we want on installing Docker itself?

Isn't that something we should just link to in Docker's official docs?

**Comment by MarcoGorelli** [comment]
IMO just a link to https://docs.docker.com/get-docker/ should be enough, though I'd keep the "**Test your Docker installation**" section because if you don't get that right (e.g. because you still need `sudo` to run `docker run`) then the next part won't work

File: doc/source/development/contributing.rst [hunk 4]
```diff
@@ -157,9 +157,56 @@ Instead of manually setting up a development environment, you can use Docker to
...
 
-Even easier, you can use the DockerFile to launch a remote session with Visual Studio Code,
-a popular free IDE, using the `.devcontainer.json` file.
-See https://code.visualstudio.com/docs/remote/containers for details.
+**Installing Docker on Windows Home**
+
+First make sure that you have Windows 10 Home, version 2004 or higher. Then proceed with enabling `Windows Subsystem for Linux <https://docs.microsoft.com/en-us/windows/wsl/install-win10>`_
+and download and install the `Linux kernel update package <https://docs.microsoft.com/pl-pl/windows/wsl/wsl2-kernel>`_.
+
+Download and install `Docker for Windows <https://docs.microsoft.com/pl-pl/windows/wsl/wsl2-kernel>`_.
```
Comment (gfyoung) [inline doc/source/development/contributing.rst:165]: Wrong link

**Comment by yehoshuadimarsky** [comment]
I'm no Docker expert either, but looks good to me, thanks @fdanieluk for your work on the docs, it really makes it accessible

**Comment by WillAyd** [comment]
I agree with @gfyoung here in that we really should just link to the external docs here

File: doc/source/development/contributing.rst [hunk 5]
```diff
@@ -157,9 +157,56 @@ Instead of manually setting up a development environment, you can use Docker to
...
+**Test your Docker installation**
+
+Run the following command in your terminal::
+
+    docker run hello-world
+
+You should see a welcome message, your installation is complete! In case of problems refer to these pages with common issues on: `Windows <https://docs.docker.com/docker-for-windows/troubleshoot/>`_,
+`MacOS <https://docs.docker.com/docker-for-mac/troubleshoot/>`_.
+
+**Development inside Docker container**
```
Comment (MarcoGorelli) [inline doc/source/development/contributing.rst:200]: Could we put
```bash
docker build --tag pandas-dev:latest .
docker run -it --rm -v $PWD:/pandas-dev -w /pandas-dev pandas-dev:latest
```
somewhere in this section, and then note that, _as an alternative_, it's possible to launch a remote session within vs code using `.devcontainer.json`?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Sure - to be fair I just checked the [tensorflow "building from source" guide](https://www.tensorflow.org/install/source) and they just link to Docker's own instructions , though they do have an example of how to start a container:

> ```
> docker pull tensorflow/tensorflow:devel
> docker run -it -w /tensorflow_src -v $PWD:/mnt -e HOST_PERMS="$(id -u):$(id -g)" \
>     tensorflow/tensorflow:devel bash
> 
> git pull  # within the container, download the latest source code
> ```

Sorry @fdanieluk for conflicting reviews from my end

**Comment by MarcoGorelli** [comment]
Hi @fdanieluk - is this still active?

**Comment by MarcoGorelli** [comment]
Closing as stale, but please ping if you'd like to continue working on this

**Pull request closed without merging**